tensorboard/webapp/widgets/histogram/histogram_v2_component.ts

```diff
@@ -55,83 +77,178 @@ export class HistogramV2Component implements OnChanges {
+  private layout: Layout = {
+    histogramHeight: 1,
+    contentClientRect: {height: 1, width: 1},
+  };
+  private scales: Scales;
+  private domInitialized = false;
+
+  constructor(private readonly changeDetector: ChangeDetectorRef) {
+    // `data` may not be available at the constructor time. Since we recalculate
+    // the scales on the `ngAfterViewInit`, let's just initialize `scales` with
+    // their default values.
+    this.scales = this.updateScales(this.layout, []);
+  }
 
   ngOnChanges() {
     if (this.content) {
-      this.ngAfterViewInit();
+      this.updateChart();
     }
   }
 
   ngAfterViewInit() {
+    this.domInitialized = true;
+    this.updateClientRects();
     this.updateChart();
+    this.changeDetector.detectChanges();
+  }
+
+  getHistogramPath(datum: HistogramDatum): string {
+    if (!this.domInitialized || !datum.bins.length) return '';
+    const xScale = this.scales.binScale;
+    const yScale = this.scales.countScale;
+
+    const firstBin = datum.bins[0];
+    const lastBin = datum.bins.slice(-1)[0];
+    const pathBuilder = [
+      `M${xScale(firstBin.x + firstBin.dx * 0.5)},${yScale(0)}`,
+    ];
+
+    for (const bin of datum.bins) {
+      pathBuilder.push(`L${xScale(bin.x + bin.dx * 0.5)},${yScale(bin.y)}`);
+    }
+
+    pathBuilder.push(`L${xScale(lastBin.x + lastBin.dx * 0.5)},${yScale(0)}`);
+    pathBuilder.push('Z');
+    return pathBuilder.join('');
+  }
+
+  trackByWallTime(datum: HistogramDatum): number {
+    return datum.wallTime;
+  }
+
+  // translates container for histogram so we can have more sensible coordinate
+  // system for reasoning with the coordinate system of a histogram.
+  getGroupTransform(datum: HistogramDatum): string {
+    if (!this.domInitialized || this.mode === HistogramMode.OVERLAY) return '';
+    return `translate(0, ${this.scales.temporalScale(
+      this.getTimeValue(datum)
+    )})`;
+  }
+
+  getHistogramFill(datum: HistogramDatum): string {
+    return this.scales.d3ColorScale(this.getTimeValue(datum));
+  }
+
+  getGridTickYLocs(): number[] {
+    if (this.mode === HistogramMode.OFFSET) return [];
+    const yScale = this.scales.countScale;
+    return yScale.ticks().map((tick) => yScale(tick));
+  }
+
+  private getTimeValue(datum: HistogramDatum): number {
+    switch (this.timeProperty) {
+      case TimeProperty.WALL_TIME:
+        return datum.wallTime;
+      case TimeProperty.STEP:
+        return datum.step;
+      case TimeProperty.RELATIVE:
+        return datum.wallTime - this.data[0].wallTime;
+    }
+  }
+
+  private updateClientRects() {
+    if (this.content) {
+      this.layout.contentClientRect = this.content.nativeElement.getBoundingClientRect();
+      this.layout.histogramHeight = this.layout.contentClientRect.height / 2.5;
+    }
   }
 
   private updateChart() {
-    this.renderXAxis(this.data);
-    this.updateYScaleAndDomain(this.data);
-    this.renderYAxis();
+    this.scales = this.updateScales(this.layout, this.data);
```

Comment:
I'm inclined to suggest dropping the params to these 3 functions, since they all come from member properties that those functions already have access to.  This call to `this.updateScales()` suggests that the only class members this function requires looking at are the `layout` and `data`, however that is not the case as

- L177 directly reads `this.layout`, ignoring the param
- L204 reads `this.colorScale`, and L248 reads `this.mode`, both properties that weren't passed in